diesel_codegen/src/insertable.rs

```diff
@@ -15,19 +17,176 @@ pub fn derive_insertable(item: syn::MacroInput) -> quote::Tokens {
     let table_name = &model.table_name();
-    let lifetimes = model.generics.lifetimes;
-    let fields = model.attrs.as_slice();
-
-    quote!(impl_Insertable! {
-        (
-            struct_name = #struct_name,
-            table_name = #table_name,
-            struct_ty = #struct_ty,
-            lifetimes = (#(#lifetimes),*),
-        ),
-        fields = [#(#fields)*],
-    })
+
+    let build_expr = model.build_expr(quote!(ref));
+
+    let value_type = model.attrs.as_slice().iter().map(|a|
+        make_column_insert_value_definition(a, table_name, &struct_name));
+    let value_type = quote!((#(#value_type,)*));
+
+    let value_tuple = model.attrs.as_slice().iter().map(|a|
+        make_column_insert_value(a, table_name, &struct_name)
+    );
+    let value_tuple = quote!((#(#value_tuple,)*));
+
+    let value_fn = quote!(
+        #[allow(non_shorthand_field_patterns)]
+        fn values(self) -> Self::Values {
+            use diesel::expression::{AsExpression, Expression};
+            use diesel::insertable::ColumnInsertValue;
+            use diesel::types::IntoNullable;
+            let #build_expr = *self;
+            #value_tuple
+        });
+
+    quote!(
+        impl#generics diesel::insertable::Insertable<#table_name::table, DB>
+            for &#insert #struct_ty
+            where DB: diesel::backend::Backend,
+                  #value_type: diesel::insertable::InsertValues<DB>,
+        {
+            type Values = #value_type;
+
+            #value_fn
+        }
+    )
+}
+
+fn make_column_insert_value_definition(
+    a: &Attr,
+    table_name: &syn::Ident,
+    struct_name: &str
+) -> quote::Tokens {
+    let column_name = a.column_name
+        .as_ref()
+        .expect(&format!("Unknown column name while implementing Insertable for {}"
+                         , struct_name));
+    let field_ty = &a.ty;
+    quote!(
+        diesel::insertable::ColumnInsertValue<
+            #table_name::#column_name,
+            diesel::expression::helper_types::AsNullableExpr<
+                &'insert #field_ty,
+                #table_name::#column_name,
+            >
+        >)
+}
+
+fn make_column_insert_value(
+    a: &Attr,
+    table_name: &syn::Ident,
+    struct_name: &str
+) -> quote::Tokens {
+    let column_name = a.column_name
+        .as_ref()
+        .expect(&format!("Unknown column name while implementing Insertable for {}"
+                         , struct_name));
+    let column = quote!(#table_name::#column_name);
+    let field_access = a.name_for_pattern();
+
+    match a.field_kind() {
+        "option" => {
+            quote!(
+                match #field_access {
+                    value @ &Some(_) => {
+                        ColumnInsertValue::Expression(
+                            #column,
+                            AsExpression::<<<#column as Expression>::SqlType
+                                as IntoNullable>::Nullable>::as_expression(value)
```

Comment:
Can you try not to break the line here? I'd split this up into

```rust
type NullableColumn = AsExpression::<
    <<#column as Expression>::SqlType as IntoNullable>::Nullable
>;
```

and then use `NullableColumn::as_expression(value)`